api/v1/db/json/notes-dao.js

```diff
@@ -0,0 +1,96 @@
+const _ = require('lodash');
+const fs = require('fs');
+const config = require('config');
+
+const { serializeNotes } = require('../../serializers/notes-serializer');
+
+/**
+ * @summary Filter notes using parameters
+ * @function
+ * @param {Object[]} rawNotes The list of notes to be filtered
+ * @param {Object} queryParams Key-value pairs of query parameters and their values
+ * @returns {Object[]} List of filtered notes
+ */
+const filterNotes = (rawNotes, queryParams) => {
+  // Safely access contextType
+  const getContextType = rawNote => (rawNote.context ? rawNote.context.contextType : null);
+
+  const {
+    q, sources, sortKey, contextTypes,
+  } = queryParams;
+
+  rawNotes = contextTypes
+    ? _.filter(rawNotes, it => _.includes(contextTypes, getContextType(it)))
+    : rawNotes;
+  rawNotes = q ? _.filter(rawNotes, it => _.includes(it.note, q)) : rawNotes;
+  // sort first by sortKey, and then by lastModified within each sorted group
+  rawNotes = _.orderBy(
+    rawNotes,
+    [sortKey === 'contextType' ? it => getContextType(it) : sortKey, 'lastModified'],
+    [sortKey === 'lastModified' ? 'desc' : 'asc', 'desc'],
+  );
+  rawNotes = sources ? _.filter(rawNotes, it => _.includes(sources, it.source)) : rawNotes;
+  return rawNotes;
+};
+
+/**
+ * @summary Return a list of notes filtered/sorted by query parameters
+ * @function
+ * @param {Object} query Query parameters
+ * @returns {Promise} Promise object represents a list of notes
+ */
+const getNotes = query => new Promise((resolve, reject) => {
+  // This is the value of the 'source' field that will be set for all notes fetched from the DB.
+  const sourceValue = 'advisorPortal';
+
+  try {
+    const { dbDirectoryPath } = config.api;
+    if (!fs.existsSync(dbDirectoryPath)) {
+      reject(new Error(`DB directory path: '${dbDirectoryPath}' is invalid`));
+    }
+
+    const { studentID } = query;
+    const studentDirPath = `${dbDirectoryPath}/${studentID}`;
+    let noteFiles;
+    let rawNotes = [];
+    try {
+      noteFiles = fs.readdirSync(studentDirPath);
+    } catch (ignore) {
+      // rawNotes should remain an empty array if directory does not exist
+    }
```

Comment:
Like we discussed in person and just my two cents, the functionality works fine but I guess it'd be better to use an if-else block here to see if the student sub-folder exist or not instead of using try-catch. Using nested try-catch block can be confusing sometimes, especially when your catch block doesn't contain any logic to handle the exception.

Reply:
Makes sense to me. This also will make sure that any unexpected errors thrown during the read operation will correctly cause the Promise to be rejected instead of silently setting `rawNotes` to `[]`